### DISCUSSION TITLE
[Detector Support]: Yolo-NAS notebook from documentation fails to run with 404 error

### DISCUSSION THREAD

**Discussion by kidhasmoxy** [blakeblackshear/frigate#14457] [Detector Support] [answered]
### Describe the problem you are having

The pretrained weights for YOLO-NAS are no longer available with a 404 error. The domain they are hosted under is for sale, so any automated attempts to download files from the model zoo will fail. This makes the documentation impossible to follow for OpenVino Yolo-NAS setup. 

Deci, the company behind the super-gradients library was acquired by Nvidia earlier this year. It's unclear whether the weights will be moved to a new location or whether the library will continue with the license it had. 

### Version

14

### Frigate config file

```yaml
! pip install -q super_gradients==3.7.1
     

from super_gradients.common.object_names import Models
from super_gradients.conversion import DetectionOutputFormatMode
from super_gradients.training import models

model = models.get(Models.YOLO_NAS_S, pretrained_weights="coco")
     

# export the model for compatibility with Frigate

model.export("yolo_nas_s.onnx",
             output_predictions_format=DetectionOutputFormatMode.FLAT_FORMAT,
             max_predictions_per_image=20,
             confidence_threshold=0.4,
             input_image_shape=(320,320),
            )
     

from google.colab import files

files.download('yolo_nas_s.onnx')
```

### docker-compose file or Docker CLI command

```yaml
N/A
```

### Relevant Frigate log output

```shell
[2024-10-20 13:16:58] INFO - crash_tips_setup.py - Crash tips is enabled. You can set your environment variable to CRASH_HANDLER=FALSE to disable it
[2024-10-20 13:17:09] INFO - utils.py - NumExpr defaulting to 2 threads.
[2024-10-20 13:17:17] WARNING - checkpoint_utils.py - :warning: The pre-trained models provided by SuperGradients may have their own licenses or terms and conditions derived from the dataset used for pre-training.
 It is your responsibility to determine whether you have permission to use the models for your use case.
 The model you have requested was pre-trained on the coco dataset, published under the following terms: https://cocodataset.org/#termsofuse
[2024-10-20 13:17:17] INFO - checkpoint_utils.py - License Notification: YOLO-NAS pre-trained weights are subjected to the specific license terms and conditions detailed in 
https://github.com/Deci-AI/super-gradients/blob/master/LICENSE.YOLONAS.md
By downloading the pre-trained weight files you agree to comply with these terms.
Downloading: "https://sghub.deci.ai/models/yolo_nas_s_coco.pth" to /root/.cache/torch/hub/checkpoints/yolo_nas_s_coco.pth
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-1-c4c919ec4cd2> in <cell line: 2>()
      1 from super_gradients.training import models
----> 2 model = models.get("yolo_nas_s", pretrained_weights="coco")

11 frames
/usr/lib/python3.10/urllib/request.py in http_error_default(self, req, fp, code, msg, hdrs)
    641 class HTTPDefaultErrorHandler(BaseHandler):
    642     def http_error_default(self, req, fp, code, msg, hdrs):
--> 643         raise HTTPError(req.full_url, code, msg, hdrs, fp)
    644 
    645 class HTTPRedirectHandler(BaseHandler):

HTTPError: HTTP Error 404: Not Found
```

### Install method

HassOS Addon

### Object Detector

OpenVino

### Screenshots of the Frigate UI's System metrics pages

![image](https://github.com/user-attachments/assets/f4379171-5bea-4db4-a992-b8a811486429)

**Comment by kidhasmoxy** [comment, marked as answer]
For this running into this, there is a temporary workaround. First you need to change the first line of the notebook to import from the latest github commit. 

! pip install -U git+https://github.com/Deci-AI/super-gradients

When you run it, you'll get an error about Index out of range referencing checkpoint_utils.py. From there, you'll need to edit checkpoint_utils.py to change line 1592 from
`       unique_filename = url.split("https://sghub.deci.ai/models/")[1].replace("/", "_").replace(" ", "_")`

to

`        unique_filename = url.split("https://sg-hub-nv.s3.amazonaws.com/models/")[1].replace("/", "_").replace(" ", "_")
`

**Comment by partytimeexcellent** [comment]
I changed the first step as you suggested, but instead of an error regarding checkpoint_utils.py, I get a dependency conflict error:

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16, which is not installed.
```

**Comment by partytimeexcellent** [comment]
And when running the second step, after changing the line in checkpoint_utils.py, I still get an index error:

```
[2024-10-24 08:13:17] WARNING - checkpoint_utils.py - :warning: The pre-trained models provided by SuperGradients may have their own licenses or terms and conditions derived from the dataset used for pre-training.
 It is your responsibility to determine whether you have permission to use the models for your use case.
 The model you have requested was pre-trained on the coco dataset, published under the following terms: https://cocodataset.org/#termsofuse
[2024-10-24 08:13:17] INFO - checkpoint_utils.py - License Notification: YOLO-NAS pre-trained weights are subjected to the specific license terms and conditions detailed in 
https://github.com/Deci-AI/super-gradients/blob/master/LICENSE.YOLONAS.md
By downloading the pre-trained weight files you agree to comply with these terms.

---------------------------------------------------------------------------

IndexError                                Traceback (most recent call last)

[<ipython-input-5-c847f4ac59f2>](https://localhost:8080/#) in <cell line: 5>()
      3 from super_gradients.training import models
      4 
----> 5 model = models.get(Models.YOLO_NAS_S, pretrained_weights="coco")

3 frames

[/usr/local/lib/python3.10/dist-packages/super_gradients/training/utils/checkpoint_utils.py](https://localhost:8080/#) in load_pretrained_weights(model, architecture, pretrained_weights)
   1590         pretrained_state_dict = torch.load(url.replace("file://", ""), map_location="cpu")
   1591     else:
-> 1592         unique_filename = url.split("https://sg-hub-nv.s3.amazonaws.com/models/")[1].replace("/", "_").replace(" ", "_")
   1593         map_location = torch.device("cpu")
   1594         with wait_for_the_master(get_local_rank()):

IndexError: list index out of range
```

**Reply by dreaming0512** [reply]
you should change "url = MODEL_URLS[model_url_key]" to url = "https://sg-hub-nv.s3.amazonaws.com/models/yolo_nas_s_coco.pth" in checkpoint_utils.py.

It's on the 1571th line in checkpoint_utils.py.

**Comment by kidhasmoxy** [comment]
That usually means you need to restart the Runtime to get it to take the
changes to the environment. That's required anytime you change the cached
python files.

**Reply by partytimeexcellent** [reply]
Ok yes! this was the missing key.

For any noobs like me here is a slightly more detailed tutorial:

1. Change the first line of the notebook to import from the latest github commit:
replace: `! pip install -q super_gradients==3.7.1`
with: `! pip install -U git+https://github.com/Deci-AI/super-gradients`

2. Run Cell 1
You will probably receive the following error, (which can be ignored) and be prompted to restart the environment, which you should do.
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16, which is not installed.
```
3.  Run Cell 2
You will receive the following error:
```
IndexError                                Traceback (most recent call last)

[<ipython-input-1-c847f4ac59f2>](https://localhost:8080/#) in <cell line: 5>()
      3 from super_gradients.training import models
      4 
----> 5 model = models.get(Models.YOLO_NAS_S, pretrained_weights="coco")

3 frames

[/usr/local/lib/python3.10/dist-packages/super_gradients/training/utils/checkpoint_utils.py](https://localhost:8080/#) in load_pretrained_weights(model, architecture, pretrained_weights)
   1590         pretrained_state_dict = torch.load(url.replace("file://", ""), map_location="cpu")
   1591     else:
-> 1592         unique_filename = url.split("https://sghub.deci.ai/models/")[1].replace("/", "_").replace(" ", "_")
   1593         map_location = torch.device("cpu")
   1594         with wait_for_the_master(get_local_rank()):

IndexError: list index out of range
```
4. The error references `/usr/local/lib/python3.10/dist-packages/super_gradients/training/utils/checkpoint_utils.py`
Click on this link to open the file editor and locate line 1592

5. Replace `unique_filename = url.split("https://sghub.deci.ai/models/")[1].replace("/", "_").replace(" ", "_")`
With `unique_filename = url.split("https://sg-hub-nv.s3.amazonaws.com/models/")[1].replace("/", "_").replace(" ", "_")`

6. Navigate to "Runtime" menu and click "Restart Session"
7.  Run Cell 2 again, this time there should be no error
8. Run Cell 3
9. Run Cell 4 and save the file to local storage

I tested this on both Chrome and Firefox, working on both browsers as of 10/24/2024

**Comment by numminorih-git** [comment]
29/10/2024 work. Thx!

**Reply by numminorih-git** [reply]
upd 04/11/2024
procedure works but frigate crashes with:
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: input for the following indices
index: 1 Got: 320 Expected: 3
index: 3 Got: 3 Expected: 320
Please fix either the inputs/outputs or the model.

**Reply by NickM-27** [reply]
This means your model config is wrong

**Reply by numminorih-git** [reply]
> This means your model config is wrong

it's clear,  then which way is right?

**Reply by NickM-27** [reply]
can't make a suggestion without seeing your config

**Reply by numminorih-git** [reply]
> can't make a suggestion without seeing your config

it's "as in" config:

```
detectors:
  onnx:
    type: onnx

model:
  model_type: yolonas
  width: 320 # <--- should match whatever was set in notebook
  height: 320 # <--- should match whatever was set in notebook
  input_pixel_format: bgr
  path: /config/yolo_nas_s.onnx
  labelmap_path: /labelmap/coco-80.txt
```

With standart tensorrt - work flawlessly, with new  onnx - error that I posted above.

Thx

**Reply by NickM-27** [reply]
you are not setting the input_tensor, which is required

**Reply by numminorih-git** [reply]
> you are not setting the input_tensor, which is required

did you mean "input_tensor: nchw"?
Thx I will try it, it missed in onnx documentation  section, so I don't paste it.

**Reply by distinctjuggle** [reply]
@NickM-27 The documentation for the ONNX detector in the beta doesn't say anything about setting the input_tensor, so if it is required, the documentation should be updated to reflect that.  

For anyone else with the same problem, the correct setting seems to be: `input_tensor: nchw` underneath `model:`

**Reply by NickM-27** [reply]
yep, will be updated in next PR

**Reply by Korenchkin** [reply]
> @NickM-27 The documentation for the ONNX detector in the beta doesn't say anything about setting the input_tensor, so if it is required, the documentation should be updated to reflect that.
> 
> For anyone else with the same problem, the correct setting seems to be: `input_tensor: nchw` underneath `model:`

not only onnx,openvino too,same problem with yolonas and nchw (n?c?Height Width probably)..not that i know what i'm doing :)

**Reply by NickM-27** [reply]
That is incorrect, openvino has always had the correct config shown https://docs.frigate.video/configuration/object_detectors#yolo-nas

**Reply by Korenchkin** [reply]
> That is incorrect, openvino has always had the correct config shown https://docs.frigate.video/configuration/object_detectors#yolo-nas

Ah,sorry about that,too much info and variants of config i'm getting lost in it

**Comment by distinctjuggle** [comment]
I'm getting a completely different error here than others.  I've changed the line 1 endpoint, ran it, restarted the environment, then ran block 2:

```
---------------------------------------------------------------------------

TypeError                                 Traceback (most recent call last)

[<ipython-input-1-c847f4ac59f2>](https://localhost:8080/#) in <cell line: 1>()
----> 1 from super_gradients.common.object_names import Models
      2 from super_gradients.conversion import DetectionOutputFormatMode
      3 from super_gradients.training import models
      4 
      5 model = models.get(Models.YOLO_NAS_S, pretrained_weights="coco")

12 frames

[/usr/lib/python3.10/abc.py](https://localhost:8080/#) in __subclasscheck__(cls, subclass)
    121         def __subclasscheck__(cls, subclass):
    122             """Override for issubclass(subclass, cls)."""
--> 123             return _abc_subclasscheck(cls, subclass)
    124 
    125         def _dump_registry(cls, file=None):

TypeError: issubclass() arg 1 must be a class
```  

If I *rerun* this block 2 **with no other changes or steps being made/taken,** the error will change?

```
---------------------------------------------------------------------------

ImportError                               Traceback (most recent call last)

[<ipython-input-2-c847f4ac59f2>](https://localhost:8080/#) in <cell line: 3>()
      1 from super_gradients.common.object_names import Models
      2 from super_gradients.conversion import DetectionOutputFormatMode
----> 3 from super_gradients.training import models
      4 
      5 model = models.get(Models.YOLO_NAS_S, pretrained_weights="coco")

1 frames

[/usr/local/lib/python3.10/dist-packages/super_gradients/training/__init__.py](https://localhost:8080/#) in <module>
      1 # PACKAGE IMPORTS FOR EXTERNAL USAGE
----> 2 import super_gradients.training.utils.distributed_training_utils as distributed_training_utils
      3 from super_gradients.common import MultiGPUMode, StrictLoad, EvaluationType
      4 from super_gradients.training.datasets import datasets_utils, DataAugmentation
      5 from super_gradients.training.pre_launch_callbacks import modify_params_for_qat

ImportError: cannot import name 'utils' from partially initialized module 'super_gradients.training' (most likely due to a circular import) (/usr/local/lib/python3.10/dist-packages/super_gradients/training/__init__.py)


---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
---------------------------------------------------------------------------
```

I've tried resetting the environment by Runtime -> Disconnect and Delete Runtime but that doesn't seem to fix the issue.

**Reply by distinctjuggle** [reply]
@ProsjektX (tagging for your benefit)

I found a workaround using a specific commit version to get the same error as documented here:

Change the first line/block to: `pip install -U git+https://github.com/Deci-AI/super-gradients@e0ccacf8868ffa1296fa4f8407c03d2bc227312c`

Then you can follow all other steps as outlined.

**Reply by Devchili** [reply]
> pip install -U git+https://github.com/Deci-AI/super-gradients@e0ccacf8868ffa1296fa4f8407c03d2bc227312c

@distinctjuggle are you installing supervision?

**Reply by distinctjuggle** [reply]
I'm not knowingly doing anything other than as described.  I don't recall supervision being anything I explicitly touched one way or the other.

**Comment by ProsjektX** [comment]
The procedure by partytimeexcellent worked a few days ago, but now I get the same fault as in the previous post. Have no clue though..

**Comment by jurasjo** [comment]
03/11/2024 procedure works.

**Reply by distinctjuggle** [reply]
What steps are you following?  Because I'm still getting the same error that I posted yesterday, which is not resolved by any procedure posted here.

**Reply by jurasjo** [reply]
Those posted by partytimeexcellent

**Reply by distinctjuggle** [reply]
Can't reproduce your success.  I'm still getting the same error as I posted a few days ago.

**Reply by jerome71** [reply]
I had to change in cell 2 python3.10 to python3.11 (the logs of cell 1 showed python3.11 being used) for it to work.
After that, success! :-)

**Comment by saket424** [comment]
@NickM-27 

Thanks for the rocm build and the yolo-nas notebook instructions
![Screenshot from 2024-11-04 16-26-54](https://github.com/user-attachments/assets/fd066fe2-7ac7-45d2-9814-e42421ba8641)

I am having the opposite problem now, My amd rocm gpu usage is high and the cpu usage is low and I am only doing 2 cameras

**Reply by NickM-27** [reply]
What hardware are you running on?

**Comment by saket424** [comment]
model name	: AMD Ryzen 5 8640U w/ Radeon 760M Graphics
https://www.newegg.com/asrock-industrial-4x4-box-8640u-amd-ryzen-5-8640u/p/N82E16856179011

**Reply by NickM-27** [reply]
this is why the docs specifically say `discrete GPU` for AMD, because integrated ROCm GPUs do not run the yolonas models efficiently

**Comment by saket424** [comment]
the amd-radeonsi-vaapi is taking a bigger share of the gpu% compared to the rocm detection

**Reply by NickM-27** [reply]
where do you see that? by the way you should create your own discussion, this is not the right place to discuss this

**Comment by victpork** [comment]
Hi everyone,
Stumbled upon this thread when trying to generate a Yolo-NAS model. In case anyone still struggling to generate the model, here is what I got after trying the methods above with no success:

1. Open the original notebook bundled with Frigate in [Google Colab](https://colab.research.google.com/github/blakeblackshear/frigate/blob/dev/notebooks/YOLO_NAS_Pretrained_Export.ipynb) 
2. Add a new code cell after the first one, and paste the code below into it: 
```
! sed -i 's/sghub.deci.ai/sg-hub-nv.s3.amazonaws.com/' /usr/local/lib/python3.10/dist-packages/super_gradients/training/pretrained_models.py
! sed -i 's/sghub.deci.ai/sg-hub-nv.s3.amazonaws.com/' /usr/local/lib/python3.10/dist-packages/super_gradients/training/utils/checkpoint_utils.py
```
![image](https://github.com/user-attachments/assets/916c71f7-f04d-4c2d-9c34-9a17b120f83e)

3. Run the code blocks one by one in sequence.

**Reply by sakcaj** [reply]
@blakeblackshear can you please update the notebook linked under https://docs.frigate.video/configuration/object_detectors#yolo-nas to the above? I've struggled with the this for ~1h, finally found this thread and the above worked flawlessly.

**Reply by victpork** [reply]
I created PR #15087 for the change, cheers :)

**Reply by CvH** [reply]
here is the working link for the notebook https://colab.research.google.com/github/blakeblackshear/frigate/blob/master/notebooks/YOLO_NAS_Pretrained_Export.ipynb

The current link at the official docs is still targeted to the dev branch where these pr was never merged/pushed to.

**Comment by midwesternrodent** [comment]
Hey, for anyone finding this discussion on google before the PR with victpork's changes is out of master. Follow [their comment](https://github.com/blakeblackshear/frigate/discussions/14457#discussioncomment-11281027) first.

If you've tried other solutions here before scrolling down this far and are now getting a bunch of errors about something being wrong in `abc.py` or are having issues importing modules due to circular imports, and you're a newbie to google colab (like me). EVEN IF, you've closed the tab, reopened it, reset the runtime, etc... unless you specifically go to runtime > disconnect and delete runtime there still might be some residual things waiting around from a previous run and causing all sorts of issues for you.

Disconnect and delete your runtime, open the notebook again, follow [victpork's suggestion](https://github.com/blakeblackshear/frigate/discussions/14457#discussioncomment-11281027) and it should work just fine.

![Screenshot_20250105_105010](https://github.com/user-attachments/assets/3a72c326-d598-4e3d-a7d9-90417b0e5c70)

**Reply by rrowe629** [reply]
No doubt it worked 2 weeks ago, but today (1/17)...

ERROR: Cannot install super-gradients==3.7.1 and super_gradients because these package versions have conflicting dependencies.

The conflict is caused by:
    super-gradients 3.7.1 depends on numpy<=1.23
    onnx 1.15.0 depends on numpy
    onnxruntime 1.15.0 depends on numpy>=1.24.2

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip to attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

**Reply by CvH** [reply]
just run https://colab.research.google.com/github/blakeblackshear/frigate/blob/master/notebooks/YOLO_NAS_Pretrained_Export.ipynb
works like a charm

**Reply by rrowe629** [reply]
Thanks for the quick reply. That is the link I was using. Unfortunately, it’s still not working for me. I have tried Disconnect and Delete. I get the same result regarding the conflict described above when running the first cell. Perhaps the requirements for onnxruntime have changed recently? Is it still working for you?

**Reply by CvH** [reply]
I ran it 3 days ago without any problem :D isn't working today anymore .. 
Maybe [this link](https://we.tl/t-S3KLLv1CpS) is working

**Comment by rrowe629** [comment]
Yes; that link worked. Thank you.

**Comment by ricopicouk** [comment]
Hi Everyone, 

I have managed to get this working - today's date is 23/01/2025.

<thanks to [victpork](https://github.com/victpork) for his code in replacing the URL> 


1.  Open the original notebook bundled with Frigate in [Google Colab](https://colab.research.google.com/github/blakeblackshear/frigate/blob/dev/notebooks/YOLO_NAS_Pretrained_Export.ipynb)
2. Replace the first line of code with 
`pip install git+https://github.com/Deci-AI/super-gradients.git
`

This will install the super-gradients from git directly. 

My colab complained about some files being uninstalled. It asked to reboot the kernel and  - I did so, and then ran the command again and it confirmed to be all installed. 

Add a new code cell after the first one, and paste the code below into it:
```
! sed -i 's/sghub.deci.ai/sg-hub-nv.s3.amazonaws.com/' /usr/local/lib/python3.11/dist-packages/super_gradients/training/pretrained_models.py
! sed -i 's/sghub.deci.ai/sg-hub-nv.s3.amazonaws.com/' /usr/local/lib/python3.11/dist-packages/super_gradients/training/utils/checkpoint_utils.py
```


Run the code following the install one by one in sequence.

**Reply by davidecavestro** [reply]
so it is a matter of installing super-gradients from git AND THEN replacing refs from deci.ai to s3 into the installed files... I'm just wondering if it makes sense forking the repo and fix that ref, so that the sed is not needed.

**Comment by ross-rosario** [comment]
If anyone see this in the future, please use Python 3.11, otherwise super-gradients won't install.
